cmd/promtool/rules.go

```diff
@@ -0,0 +1,208 @@
+)
+
+const blockSize = 2 // in hours
+
+// RuleImporter is the importer to backfill rules.
+type RuleImporter struct {
+	logger log.Logger
+	config RuleImporterConfig
+
+	groups      map[string]*rules.Group
+	groupLoader rules.GroupLoader
+
+	apiClient v1.API
+
+	writer *blocks.MultiWriter
+}
+
+// RuleImporterConfig is the config for the rule importer.
+type RuleImporterConfig struct {
+	Start        time.Time
+	End          time.Time
+	OutputDir    string
+	EvalInterval time.Duration
+	URL          string
+}
+
+// NewRuleImporter creates a new rule importer that can be used to backfill rules.
+func NewRuleImporter(logger log.Logger, config RuleImporterConfig) *RuleImporter {
+	return &RuleImporter{
+		config:      config,
+		groupLoader: rules.FileLoader{},
+	}
+}
+
+// Init initializes the rule importer which creates a new block writer
+// and creates an Prometheus API client.
+func (importer *RuleImporter) Init() error {
+	importer.writer = blocks.NewMultiWriter(importer.logger,
+		importer.config.OutputDir,
+		importer.config.EvalInterval.Nanoseconds(),
+	)
+
+	config := api.Config{
+		Address: importer.config.URL,
+	}
+	c, err := api.NewClient(config)
+	if err != nil {
+		return err
+	}
+	importer.apiClient = v1.NewAPI(c)
+	return nil
+}
+
+// Close cleans up any open resources.
+func (importer *RuleImporter) Close() error {
+	return importer.writer.Close()
+}
+
+// LoadGroups reads groups from a list of rule files.
+func (importer *RuleImporter) LoadGroups(ctx context.Context, filenames []string) (errs []error) {
+	groups := make(map[string]*rules.Group)
+
+	for _, filename := range filenames {
+		rgs, errs := importer.groupLoader.Load(filename)
+		if errs != nil {
+			return errs
+		}
+
+		for _, ruleGroup := range rgs.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rgRules := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := importer.groupLoader.Parse(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, filename)}
+				}
+				rgRules = append(rgRules, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[rules.GroupKey(filename, ruleGroup.Name)] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     filename,
+				Interval: itv,
+				Rules:    rgRules,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return nil
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores them in new blocks.
+func (importer *RuleImporter) ImportAll(ctx context.Context) []error {
+	var errs = []error{}
+	for _, group := range importer.groups {
+		stimeWithAlignment := group.EvalTimestamp(importer.config.Start.UnixNano())
+
+		for _, r := range group.Rules() {
+			err := importer.ImportRule(ctx, r.Query().String(), stimeWithAlignment, group.Interval())
```

Comment:
I think we should be going block by block, rather than rule by rule. It should be more efficient.

Reply:
good idea. Let me change that

Reply:
Do we mean having single rule (series) per block vs multiple per block? Let's dive into this more before deciding, I am not sure if this is more efficient. Spamming one-series blocks vs having one block with X series for X rules seems not very efficial me, but both of this will "work" for now. Why single series per block would be more efficient?